### PR TITLE
Support CLI tools that use phars without a .phar extension

### PR DESCRIPTION
**Pull request by alexpott**
Some CLI tools are shipped as phar files and the install instructions tell users to rename the cli_command.phar to cli_command. 

In this instance the interceptor can produce false positives because it is handling a phar file without a phar extension. But the very first caller is a phar file so I think it is safe to assume this is intentional.

### CODE REVIEW

**Comment by alexpott** [comment]
For example of such a CLI tool - see https://getcomposer.org/doc/00-intro.md#globally

**Comment by alexpott** [comment]
Here's the code for cli_phar as it is a binary diff and therefore is not so easy to work with.

`
<?php
// Use the current working directory so we don't have to include all the code in
// the phar file.
require_once getcwd() . '/../../../vendor/autoload.php';
stream_wrapper_unregister('phar');
stream_wrapper_register('phar', \TYPO3\PharStreamWrapper\PharStreamWrapper::class);

\TYPO3\PharStreamWrapper\Manager::initialize(
    (new \TYPO3\PharStreamWrapper\Behavior())
        ->withAssertion(new \TYPO3\PharStreamWrapper\Interceptor\PharExtensionInterceptor())
);

// This should not throw an exception.
if (file_exists(__DIR__ . '/index.php')) {
  echo "Can access phar files without .phar extension if they are the CLI command.\n";
}

// Try an insecure phar without an extension.
if (file_exists('phar://bundle.phar')) {
  echo "Can access phar files with .phar extension.\n";
}

try {
  file_exists('phar://bundle.phar.png');
}
catch (\TYPO3\PharStreamWrapper\Exception $e) {
  echo "Cannot access other phar files without .phar extension.\n";
}
`

**Comment by ohader** [comment]
Thanks for your pull-request. I'm going to look into that in more details during the next days...

**Comment by ohader** [comment]
In addition I've created an experimental Phar meta-data analyzer a couple of weeks ago in branch https://github.com/TYPO3/phar-stream-wrapper/tree/feature/phar-reader

```
        Manager::initialize(
            (new \TYPO3\PharStreamWrapper\Behavior())
                ->withAssertion(new PharMetaDataInterceptor())
        );
```

This way if could be detected, whether the file itself is malicious in terms of insecure deserialization - istead of just checking the file extension. Any feedback or tests are welcome of course...

**Comment by mortenson** [comment]
Note: This PR will need updated - per https://www.drupal.org/project/drupal/issues/3026386, not all backtraces have a "file" key, and some "file" entries include the `phar://` extension.

**Comment by alexpott** [comment]
Drupal discovered another issue now we've put this out in a security release - see https://www.drupal.org/project/drupal/issues/3026443

Basically Phar aliases break \TYPO3\PharStreamWrapper\Helper::determineBaseFile() because there's no way that can resolve the Phar alias that only exists inside the phar file.

**Comment by ohader** [comment]
Thanks for your feedback. I tried to read through the comments in the issue you mentioned, but failed to understand the according context. Are those failure related to CLI invocation only? Or does it happen when being called in a web context (using different PHP_SAPI)?

If you could point me to source code or 3rd party packages and how they are using Phar aliases, I could create according test cases.

**Comment by ohader** [comment]
@alexpott Just read your last comment https://www.drupal.org/project/drupal/issues/3026443#comment-12930646

```
$phar = new \Phar(__DIR__ . '/alias_phar.phar', NULL, 'alias_phar.phar');
var_dump($phar->getAlias());```

I assume that should be used in order to determine the alias (I still have to dig into the problem in real source code)... However, having `new \Phar($baseName)` will already execute meta-data deserialization - which might be insecure and compromised.

**Comment by alexpott** [comment]
@ohader re `which might be insecure and compromised.` yeah let's why we only do this after checking the extension. So basically we're at the point where we would allow access. I think the this library what we need to do is store a phars alias once we've determined it to be safe.

**Comment by alexpott** [comment]
@ohader the drupal issue has all the test code, necessary test phars and also the code to build them :)

File: src/PharStreamWrapper.php [hunk 1]
```diff
@@ -411,7 +418,28 @@ public function url_stat(string $path, int $flags)
...
+            foreach ($this->aliases as $alias) {
+                if (strpos($path, 'phar://' . $alias) === 0) {
+                    return;
+                }
+            }
+        }
         if ($this->resolveAssertable()->assert($path, $command) === true) {
+            if ($basePath && !isset($this->aliases[$path])) {
+                $this->restoreInternalSteamWrapper();
+                $this->aliases[$path] = (new \Phar(realpath($basePath)))->getAlias();
```
Comment (LionsAd) [inline src/PharStreamWrapper.php]: $path should be $basePath here unless I miss something.

**Comment by alexpott** [comment]
So there are some issues with this PR.
* The architecture of the stream wrapper allows for per command interception - the new cache / alias approach breaks that. We don't have test coverage of this. I wonder why the per command stuff was added?
* We know that not all phar commands have aliases set the same way - for example https://github.com/maxmind/GeoIP2-php/releases/download/v2.9.0/geoip2.phar - as a result of being built with box

**Comment by ohader** [comment]
* Defining behavior based on commands would have allowed different strategies for invoking interceptors - I currently don't see how this might break, given that the `$alias` cache is located in the base `PharStreamWrapper`.
* Concerning alias resolving (especially for those that are not part of Phar manifests - but using `Phar::mapPhar('alias')`) I've extended the experimental branch to include that https://github.com/TYPO3/phar-stream-wrapper/blob/feature/phar-reader/tests/Functional/Phar/ReaderTest.php - reading stub content feels ugly, but I did not find any other reference on PHP is exposing that internal Phar alias mapping...

**Comment by LionsAd** [comment]
@ohader I agree, I also searched the PHP code base for how mapPhar and setAlias work and the best I could determine was that if you use ```new Phar($filename, ..., $alias)``` -- it is only a temporary alias and hence not written to disk.

The better way for box and co would be to call ```->setAlias()```, which does write to disk and is permanent to make things easier for us.

In theory we could create a patcher tool to allow people to patch their phar files without aliases to have the same as internally used via mapPhar().

**Comment by ohader** [comment]
Alright, I'd like to proceed as following
* integrate Phar Reader feature into master
* separate this PR into CLI and alias related topics
* define & create test cases

**Comment by ohader** [comment]
Could not resolve CLI flaws anymore after having alias handling and `realpath` resolving of PR #12 applied. Find according test here https://github.com/TYPO3/phar-stream-wrapper/pull/12/commits/1ada68c323d6d06a6b21c50e5b9f316a1f6d6897

**Comment by ohader** [comment]
Seems to be resolved with c267f38a7ac4e17a81aa2edf9c7e2f6df072f240 and 42730ed6b8de6e59764e2340f3561b093f78291d

**Comment by ohader** [comment]
Address in PR #12

**Pull request closed without merging**